go-controller/pkg/node/route_manager.go

```diff
@@ -0,0 +1,483 @@
+	}
+	return nil
+}
+
+func (rm *routeManager) netlinkDelRoute(link netlink.Link, subnet *net.IPNet) error {
+	// List routes for the link in the default routing table
+	nlRoutes, err := netlink.RouteList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("failed to get routes for link %s: %v", link.Attrs().Name, err)
+	}
+	for _, nlRoute := range nlRoutes {
+		deleteRoute := false
+		if subnet == nil {
+			deleteRoute = nlRoute.Dst == nil
+		} else if nlRoute.Dst != nil {
+			deleteRoute = nlRoute.Dst.String() == subnet.String()
+		}
+
+		if deleteRoute {
+			err = netlink.RouteDel(&nlRoute)
+			if err != nil {
+				net := "default"
+				if nlRoute.Dst != nil {
+					net = nlRoute.Dst.String()
+				}
+				return fmt.Errorf("failed to delete route '%s via %s' for link %s : %v\n",
+					net, nlRoute.Gw.String(), link.Attrs().Name, err)
+			}
+			break
+		}
+	}
+	return nil
+}
+
+func (rm *routeManager) addRouteToStore(r route) error {
+	infName, err := r.getLinkName()
+	if err != nil {
+		return fmt.Errorf("failed to add route (%s) to store because we could not get link name", r.String())
+	}
+	existingR, ok := rm.store[infName]
+	if !ok {
+		rm.store[infName] = r
+		return nil
+	}
+	newRes := make([]routeEntry, 0)
+	for _, newRe := range r.routeEntries {
+		var found bool
+		for _, existingRe := range existingR.routeEntries {
+			if existingRe.equal(newRe) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			newRes = append(newRes, newRe)
+		}
+	}
+	if len(newRes) == 0 {
+		klog.Infof("Route Manager: nothing to process for new route as it is already managed: %s", r.String())
+		return nil
+	}
+	existingR.routeEntries = append(existingR.routeEntries, newRes...)
+	rm.store[infName] = existingR
+	return nil
+}
+
+func (rm *routeManager) sync() {
+	for infName, r := range rm.store {
+		activeNlRoutes, err := netlink.RouteList(r.link, nl.FAMILY_ALL)
+		if err != nil {
+			klog.Errorf("Route Manager: failed to list routes for link %q: %w", infName, err)
+			continue
+		}
+		var activeRes []routeEntry
+		for _, activeNlRoute := range activeNlRoutes {
+			activeRoute, err := convertNetlinkRouteToRoute(activeNlRoute)
+			if err != nil {
+				klog.Errorf("Route Manager: failed to convert netlink route (%s) to route: %v",
+					activeRoute.String(), err)
+				continue
+			}
+			activeRes = append(activeRes, activeRoute.routeEntries...)
+		}
+		var syncNeeded bool
+		var syncReason string
+		for _, expectedRe := range r.routeEntries {
+			var found bool
+			for _, activeRe := range activeRes {
+				if activeRe.equal(expectedRe) {
+					found = true
+				}
+			}
+			if !found {
+				syncReason = fmt.Sprintf("failed to find route: %s", expectedRe.string())
+				syncNeeded = true
+				break
+			}
+		}
+		if syncNeeded {
+			klog.Infof("Route Manager: sync required for routes associated with link %q. Reason: %s", infName, syncReason)
+			if err = rm.applyRoute(r); err != nil {
```

Comment:
shouldn't this function be applyRoutes? really you are applying all the routes stored in that link right?

Reply:
yep - renamed.